sklearn/tree/_criterion.pyx

```diff
@@ -1173,17 +1174,180 @@ cdef class MSE(RegressionCriterion):
+    end : intp_t
+        End index (exclusive) in `sample_indices`
+    abs_errors : float64_t[::1]
+        array to store (increment) the computed absolute errors. Shape: (n,)
+        with n := end - start
+    medians : float64_t[::1]
+        array to store (overwrite) the computed medians. Shape: (n,)
+
+    Complexity: O(n log n)
+    This algorithm is an adaptation of the two heaps solution of
+    the "find median from a data stream" problem
+    See for instance: https://www.geeksforgeeks.org/dsa/median-of-stream-of-integers-running-integers/
+
+    But here, it's the weighted median and we also need to compute the AE, so:
+    - instead of balancing the heaps based on their number of elements,
+      rebalance them based on the summed weights of their elements
+    - rewrite the AE computation by splitting the sum between elements
+      above and below the median, which allow to express it as a simple
+      O(1) computation.
+      See the maths in the PR desc:
+      https://github.com/scikit-learn/scikit-learn/pull/32100
+    """
+    cdef intp_t j, p, i, step, n
+    if start < end:
+        j = 0
+        step = 1
+        n = end - start
+    else:
+        n = start - end
+        step = -1
+        j = n - 1
+
+    above.reset()
+    below.reset()
+    cdef float64_t y
+    cdef float64_t w = 1.0
+    cdef float64_t top_val, top_weight
+    cdef float64_t median = 0.0
+    cdef float64_t half_weight
+
+    p = start
+    for _ in range(n):
+        i = sample_indices[p]
+        if sample_weight is not None:
+            w = sample_weight[i]
+        y = ys[i, k]
+
+        # Insert into the appropriate heap
+        if below.is_empty():
+            above.push(y, w)
+        elif y > below.top():
+            above.push(y, w)
+        else:
+            below.push(y, w)
+
+        half_weight = (above.total_weight + below.total_weight) / 2.0
+
+        # Rebalance heaps
+        while above.total_weight < half_weight and not below.is_empty():
+            below.pop(&top_val, &top_weight)
+            above.push(top_val, top_weight)
+        while (
+            not above.is_empty()
+            and (above.total_weight - above.top_weight()) >= half_weight
+        ):
+            above.pop(&top_val, &top_weight)
+            below.push(top_val, top_weight)
+
+        # Current median
+        if above.total_weight > half_weight + 1e-5 * fabs(half_weight):
+            median = above.top()
+        else:  # above and below weight are almost exactly equals
+            median = (above.top() + below.top()) / 2.
+        medians[j] = median
+        abs_errors[j] += (
+            (below.total_weight - above.total_weight) * median
+            - below.weighted_sum
+            + above.weighted_sum
+        )
+        p += step
+        j += step
+
+
+def _py_precompute_absolute_errors(
+    const float64_t[:, ::1] ys,
+    const float64_t[:] sample_weight,
+    const intp_t[:] sample_indices,
+    bint suffix=False
```

Comment:
Given this is only used for testing, can we pass in `start` and `end` directly and not use `suffix`?